embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/internal/spans/SpanSinkImpl.kt

```diff
@@ -19,16 +21,17 @@ internal class SpanSinkImpl : SpanSink {
     }
 
     override fun completedSpans(): List<EmbraceSpanData> {
-        synchronized(completedSpans) {
-            return completedSpans.toList()
-        }
+        val spansToReturn = completedSpans.size
+        return completedSpans.take(spansToReturn)
     }
 
     override fun flushSpans(): List<EmbraceSpanData> {
-        synchronized(completedSpans) {
-            val flushedSpans = completedSpans.toList()
-            completedSpans.clear()
-            return flushedSpans
+        synchronized(spansToFlush) {
+            spansToFlush.set(completedSpans())
+            repeat(spansToFlush.get().size) {
+                completedSpans.removeAll(spansToFlush.get().toSet())
```

Comment:
Is this correct? removing all as many times as spans in the list?

Reply:
hahaha this is ABSOLUTELY not correct - I will change.

I had a previous implementation that polled N times to remove and add to the flushed spans, but that leaves the the object in a weird state if it crashes in the middle. Doing this atomically works much better even if I have to convert this to a set first.